python/src/wslink/protocol.py

```diff
@@ -253,7 +255,9 @@ async def handleSystemMessage(self, rpcid, methodName, args, client_id):
 
     async def onMessage(self, is_binary, msg, client_id):
         if not is_binary:
-            logger.critical("wslink is not expecting text message:\n> %s", msg.data)
+            error_message = "wslink is not expecting text message:\n> %s"
+            logger.critical(error_message, msg.data)
+            self.log_emitter("critical", error_message % msg.data)
```

Comment:
Should that be `self.log_emitter.emit("critical", error_message % msg.data)` ?

Reply:
I had added an implementation to `__call__()` so that the shorthand notation can be used

```python
    def __call__(self, event, *args, **kwargs):
        self.emit(event, *args, **kwargs)
```

Reply:
while keeping the emit/__call__ available for out of category or programatic emit?

Reply:
The class doesn't enforce a runtime check on the event types, it's only statically typed to guide the programmer. This is why we can use it both in this case, as well as the jupyter protocol.

But yes we can have a dynamic `__getattr__` implementation that returns a partial-ed `emit()` and it would achieve what you are suggesting.

Reply:
This is the implementation:
```python
def __getattr__(self, name):
    return functools.partial(self.emit, name)
```

I would still prefer to use `emit/__call__` to have the extra safeguard when writing the code, but I don't think it hurts to have it there

Reply:
Actually, the typechecker is smart enough to catch programming errors if the signature of the function is defined as:

```python
class EventEmitter(Generic[T]):
    def __getattr__(self, name: T):
        return functools.partial(self.emit, name)

emitter = EventEmitter[Literal["debug", "error"]]()

emitter.debug() # <-- ok!
emitter.blah() # <-- type check error, would work at runtime
```
So I don't see a downside on using it

Reply:
if we get some error that we have such typo, that is great